databricks/koalas/spark.py

```diff
@@ -0,0 +1,731 @@
+
+import pyspark
+from pyspark import StorageLevel
+from pyspark.sql import Column
+
+if TYPE_CHECKING:
+    import databricks.koalas as ks
+    from databricks.koalas.base import IndexOpsMixin
+    from databricks.koalas.frame import CachedDataFrame
+
+
+class SparkIndexOpsMethods(object):
+    """Spark related features. Usually, the features here are missing in pandas
+    but Spark has it."""
+
+    def __init__(self, data: Union["IndexOpsMixin"]):
+        self._data = data
+
+    @property
+    def type(self):
+        """ Returns the data type as defined by Spark, as a Spark DataType object."""
+        return self._data._internal.spark_type_for(self._data._internal.column_labels[0])
+
+    @property
+    def column(self):
+        """
+        Spark Column object representing the Series/Index.
+
+        .. note:: This Spark Column object is strictly stick to its base DataFrame the Series/Index
+            was derived from.
+        """
+        return self._data._internal.spark_column
+
+    def transform(self, func):
+        """
+        Applies a function that takes and returns a Spark column. It allows natively
+        apply a Spark function and column APIs with the Spark column internally used
+        in Series or Index.
+
+        .. note:: It requires to have the same input and output length; therefore,
+            the aggregate Spark functions such as count does not work.
+
+        Parameters
+        ----------
+        func : function
+            Function to use for transforming the data by using Spark columns.
+
+        Returns
+        -------
+        Series or Index
+
+        Raises
+        ------
+        ValueError : If the output from the function is not a Spark column.
+
+        Examples
+        --------
+        >>> from pyspark.sql.functions import log
+        >>> df = ks.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}, columns=["a", "b"])
+        >>> df
+           a  b
+        0  1  4
+        1  2  5
+        2  3  6
+
+        >>> df.a.spark.transform(lambda c: log(c))
+        0    0.000000
+        1    0.693147
+        2    1.098612
+        Name: a, dtype: float64
+
+        >>> df.index.spark.transform(lambda c: c + 10)
+        Int64Index([10, 11, 12], dtype='int64')
+
+        >>> df.a.spark.transform(lambda c: c + df.b.spark.column)
+        0    5
+        1    7
+        2    9
+        Name: a, dtype: int64
+        """
+        from databricks.koalas import MultiIndex
+
+        if isinstance(self._data, MultiIndex):
+            raise NotImplementedError("MultiIndex does not support spark.transform yet.")
+        output = func(self._data.spark.column)
+        if not isinstance(output, Column):
+            raise ValueError(
+                "The output of the function [%s] should be of a "
+                "pyspark.sql.Column; however, got [%s]." % (func, type(output))
+            )
+        return self._data._with_new_scol(scol=func(self._data.spark.column)).rename(self._data.name)
```

Comment:
What if the output causes `AnalysisException`? E.g., 

```py
df1.a.spark.transform(lambda c: c + df2.b.spark.column)
```

Maybe we should raise an error here?